File: WorkoutTracker/Services/CSVParser.swift
import Foundation

struct CSVParser {
    enum CSVError: LocalizedError {
        case invalidHeader(String)
        case invalidColumnCount(line: Int, expected: Int, got: Int)
        case invalidField(field: String, line: Int, value: String)
        case emptyFile
        case fileNotFound(String)

        var errorDescription: String? {
            switch self {
            case .invalidHeader(let msg): return "Invalid CSV header: \(msg)"
            case .invalidColumnCount(let line, let expected, let got):
                return "Line \(line): expected \(expected) columns, got \(got)"
            case .invalidField(let field, let line, let value):
                return "Line \(line): invalid value '\(value)' for field '\(field)'"
            case .emptyFile: return "CSV file is empty"
            case .fileNotFound(let name): return "File not found: \(name)"
            }
        }
    }

    static let expectedHeaders = [
        "program_name", "week", "day_of_week", "phase", "exercise_order",
        "exercise_name", "equipment_type", "equipment_detail", "base_weight",
        "sets", "rep_min", "rep_max", "rest_seconds", "progression_rule",
        "unilateral", "notes"
    ]

    // MARK: - Public API

    static func loadBundled() -> [Program] {
        var programs: [Program] = []
        let csvFiles = ["reacher_build_workout", "daily_mobility"]

        for fileName in csvFiles {
            guard let url = Bundle.main.url(forResource: fileName, withExtension: "csv") else {
                print("CSVParser: Could not find \(fileName).csv in bundle")
                continue
            }
            do {
                let content = try String(contentsOf: url, encoding: .utf8)
                let parsed = try parse(csvString: content)
                programs.append(contentsOf: parsed)
            } catch {
                print("CSVParser: Error parsing \(fileName).csv: \(error.localizedDescription)")
            }
        }
        return programs
    }

    static func loadFromURL(_ url: URL) throws -> [Program] {
        let accessing = url.startAccessingSecurityScopedResource()
        defer {
            if accessing { url.stopAccessingSecurityScopedResource() }
        }
        let content = try String(contentsOf: url, encoding: .utf8)
        return try parse(csvString: content)
    }

    // MARK: - Parsing

    static func parse(csvString: String) throws -> [Program] {
        let lines = csvString.components(separatedBy: .newlines)
            .map { $0.trimmingCharacters(in: .whitespacesAndNewlines) }
            .filter { !$0.isEmpty }

        guard !lines.isEmpty else { throw CSVError.emptyFile }

        // Validate header
        let headerColumns = parseCSVLine(lines[0])
        let normalizedHeaders = headerColumns.map { $0.trimmingCharacters(in: .whitespaces).lowercased() }
        for (i, expected) in expectedHeaders.enumerated() {
            guard i < normalizedHeaders.count, normalizedHeaders[i] == expected else {
                throw CSVError.invalidHeader("Expected '\(expected)' at column \(i + 1)")
            }
        }

        // Parse rows
        struct RawRow {
            let programName: String
            let week: Int
            let dayOfWeek: String
            let exercise: Exercise
        }

        var rows: [RawRow] = []
        for (lineIndex, line) in lines.dropFirst().enumerated() {
            let lineNumber = lineIndex + 2
            let columns = parseCSVLine(line)

            // Allow trailing empty columns
            guard columns.count >= expectedHeaders.count else {
                throw CSVError.invalidColumnCount(line: lineNumber, expected: expectedHeaders.count, got: columns.count)
            }

            let row = try parseRow(columns: columns, lineNumber: lineNumber)
            rows.append(row)
        }

        // Group into Programs
        return buildPrograms(from: rows)
    }

    // MARK: - Row Parsing

    private static func parseRow(columns: [String], lineNumber: Int) throws -> (programName: String, week: Int, dayOfWeek: String, exercise: Exercise) {
        let col = columns.map { $0.trimmingCharacters(in: .whitespaces) }

        let programName = col[0]
        guard let week = Int(col[1]) else {
            throw CSVError.invalidField(field: "week", line: lineNumber, value: col[1])
        }
        let dayOfWeek = col[2]

        guard let phase = Phase(rawValue: col[3].lowercased()) else {
            throw CSVError.invalidField(field: "phase", line: lineNumber, value: col[3])
        }

        guard let order = Int(col[4]) else {
            throw CSVError.invalidField(field: "exercise_order", line: lineNumber, value: col[4])
        }

        let name = col[5]

        guard let equipmentType = EquipmentType(rawValue: col[6]) else {
            throw CSVError.invalidField(field: "equipment_type", line: lineNumber, value: col[6])
        }

        let equipmentDetail: String? = col[7].isEmpty ? nil : col[7]
        let baseWeight = WeightSpec(csvValue: col[8])

        guard let sets = Int(col[9]) else {
            throw CSVError.invalidField(field: "sets", line: lineNumber, value: col[9])
        }

        guard let repMin = Int(col[10]) else {
            throw CSVError.invalidField(field: "rep_min", line: lineNumber, value: col[10])
        }

        let repMax = RepTarget(csvValue: col[11])

        guard let restSeconds = Int(col[12]) else {
            throw CSVError.invalidField(field: "rest_seconds", line: lineNumber, value: col[12])
        }

        let progressionRule: ProgressionRule
        let ruleStr = col[13]
        if let rule = ProgressionRule(rawValue: ruleStr) {
            progressionRule = rule
        } else {
            // Try common alternatives
            progressionRule = .none
        }

        let unilateral = col[14].uppercased() == "TRUE"

        let notes: String? = col.count > 15 && !col[15].isEmpty ? col[15] : nil

        let exercise = Exercise(
            id: UUID(),
            order: order,
            name: name,
            phase: phase,
            equipmentType: equipmentType,
            equipmentDetail: equipmentDetail,
            baseWeight: baseWeight,
            sets: sets,
            repMin: repMin,
            repMax: repMax,
            restSeconds: restSeconds,
            progressionRule: progressionRule,
            isUnilateral: unilateral,
            notes: notes
        )

        return (programName, week, dayOfWeek, exercise)
    }

    // MARK: - CSV Line Parsing (handles commas in quoted fields)

    private static func parseCSVLine(_ line: String) -> [String] {
        var result: [String] = []
        var current = ""
        var inQuotes = false

        for char in line {
            if char == "\"" {
                inQuotes.toggle()
            } else if char == "," && !inQuotes {
                result.append(current)
                current = ""
            } else {
                current.append(char)
            }
        }
        result.append(current)
        return result
    }

    // MARK: - Build Program Hierarchy

    private static func buildPrograms(from rows: [(programName: String, week: Int, dayOfWeek: String, exercise: Exercise)]) -> [Program] {
        // Group by program name
        var programGroups: [String: [(week: Int, day: String, exercise: Exercise)]] = [:]
        for row in rows {
            programGroups[row.programName, default: []].append((row.week, row.dayOfWeek, row.exercise))
        }

        var programs: [Program] = []

        for (programName, entries) in programGroups {
            // Group by week -> day -> exercises
            var workoutsByWeekAndDay: [Int: [String: Workout]] = [:]
            var maxWeek = 1

            // Group entries by (week, day)
            var grouped: [Int: [String: [Exercise]]] = [:]
            for entry in entries {
                grouped[entry.week, default: [:]][entry.day, default: []].append(entry.exercise)
                maxWeek = max(maxWeek, entry.week)
            }

            for (week, dayExercises) in grouped {
                for (day, exercises) in dayExercises {
                    let sorted = exercises.sorted { $0.order < $1.order }
                    let workout = Workout(
                        id: UUID(),
                        programName: programName,
                        week: week,
                        dayOfWeek: day,
                        exercises: sorted
                    )
                    workoutsByWeekAndDay[week, default: [:]][day] = workout
                }
            }

            let program = Program(
                id: UUID(),
                name: programName,
                totalWeeks: maxWeek,
                currentWeek: 1,
                workoutsByWeekAndDay: workoutsByWeekAndDay
            )
            programs.append(program)
        }

        return programs.sorted { $0.name < $1.name }
    }
}

File: web/src/lib/csv-parser.ts
import type {
  Exercise, Program, Workout, Phase, EquipmentType,
  ProgressionRule, WeightSpec, RepTarget,
} from "./types";

const EXPECTED_HEADERS = [
  "program_name", "week", "day_of_week", "phase", "exercise_order",
  "exercise_name", "equipment_type", "equipment_detail", "base_weight",
  "sets", "rep_min", "rep_max", "rest_seconds", "progression_rule",
  "unilateral", "notes",
];

const VALID_PHASES = new Set<Phase>(["warmup", "main", "finisher", "cooldown", "mobility"]);
const VALID_EQUIPMENT: Set<string> = new Set([
  "barbell_45", "barbell_35", "barbell_ez", "powerblock",
  "band", "kettlebell", "bodyweight", "assisted_pullup", "gripper",
]);
// Known keyword rules — any other non-empty string is treated as a free-form progression rule
const KEYWORD_PROGRESSIONS: Set<string> = new Set([
  "add_5lb", "add_2.5lb", "add_10lb",
  "add_reps", "add_time", "add_rounds",
  "maintain", "deload", "progress_gripper", "none",
]);

function parseCSVLine(line: string): string[] {
  const result: string[] = [];
  let current = "";
  let inQuotes = false;
  for (const char of line) {
    if (char === '"') {
      inQuotes = !inQuotes;
    } else if (char === "," && !inQuotes) {
      result.push(current);
      current = "";
    } else {
      current += char;
    }
  }
  result.push(current);
  return result;
}

export function parseCSV(csvString: string): { programs: Omit<Program, "createdAt">[]; workouts: Workout[] } {
  const lines = csvString
    .split(/\r?\n/)
    .map((l) => l.trim())
    .filter((l) => l.length > 0);

  if (lines.length === 0) throw new Error("CSV file is empty");

  // Validate header
  const headerCols = parseCSVLine(lines[0]).map((h) => h.trim().toLowerCase());
  for (let i = 0; i < EXPECTED_HEADERS.length; i++) {
    if (i >= headerCols.length || headerCols[i] !== EXPECTED_HEADERS[i]) {
      throw new Error(`Invalid header: expected '${EXPECTED_HEADERS[i]}' at column ${i + 1}`);
    }
  }

  // Parse rows
  interface RawRow {
    programName: string;
    week: number;
    day: string;
    exercise: Exercise;
  }

  const rows: RawRow[] = [];

  for (let i = 1; i < lines.length; i++) {
    const cols = parseCSVLine(lines[i]).map((c) => c.trim());
    if (cols.length < EXPECTED_HEADERS.length) {
      throw new Error(`Line ${i + 1}: expected ${EXPECTED_HEADERS.length} columns, got ${cols.length}`);
    }

    const programName = cols[0];
    const week = parseInt(cols[1]);
    if (isNaN(week)) throw new Error(`Line ${i + 1}: invalid week '${cols[1]}'`);

    const day = cols[2];
    const phase = cols[3].toLowerCase() as Phase;
    if (!VALID_PHASES.has(phase)) throw new Error(`Line ${i + 1}: invalid phase '${cols[3]}'`);

    const order = parseInt(cols[4]);
    if (isNaN(order)) throw new Error(`Line ${i + 1}: invalid order '${cols[4]}'`);

    const name = cols[5];
    const equipmentType = cols[6] as EquipmentType;
    if (!VALID_EQUIPMENT.has(equipmentType)) throw new Error(`Line ${i + 1}: invalid equipment '${cols[6]}'`);

    const equipmentDetail = cols[7] || null;

    const baseWeight: WeightSpec = cols[8].toLowerCase() === "progressive"
      ? { type: "progressive" }
      : { type: "fixed", value: parseFloat(cols[8]) || 0 };

    const sets = parseInt(cols[9]) || 0;
    const repMin = parseInt(cols[10]) || 0;

    const repMax: RepTarget = cols[11].toLowerCase() === "failure"
      ? { type: "failure" }
      : { type: "count", value: parseInt(cols[11]) || 0 };

    const restSeconds = parseInt(cols[12]) || 0;

    // Known keywords are accepted as-is; any other non-empty string is a free-form rule
    const progressionRule: ProgressionRule = cols[13] || "none";

    const isUnilateral = cols[14].toUpperCase() === "TRUE";
    const notes = cols.length > 15 && cols[15] ? cols[15] : null;

    rows.push({
      programName,
      week,
      day,
      exercise: {
        id: crypto.randomUUID(),
        order,
        name,
        phase,
        equipmentType,
        equipmentDetail,
        baseWeight,
        sets,
        repMin,
        repMax,
        restSeconds,
        progressionRule,
        isUnilateral,
        isTimeBased: progressionRule === "add_time" || repMin >= 30,
        notes,
      },
    });
  }

  // Group into programs and workouts
  const programMap = new Map<string, { maxWeek: number }>();
  const workoutMap = new Map<string, { programName: string; week: number; day: string; exercises: Exercise[] }>();

  for (const row of rows) {
    const existing = programMap.get(row.programName);
    if (!existing || row.week > existing.maxWeek) {
      programMap.set(row.programName, { maxWeek: row.week });
    }

    const key = `${row.programName}_${row.week}_${row.day}`;
    const wk = workoutMap.get(key);
    if (wk) {
      wk.exercises.push(row.exercise);
    } else {
      workoutMap.set(key, {
        programName: row.programName,
        week: row.week,
        day: row.day,
        exercises: [row.exercise],
      });
    }
  }

  const programs: Omit<Program, "createdAt">[] = [];
  for (const [name, data] of programMap) {
    programs.push({
      id: name.toLowerCase().replace(/\s+/g, "-"),
      name,
      totalWeeks: data.maxWeek,
    });
  }

  const workouts: Workout[] = [];
  for (const [key, data] of workoutMap) {
    data.exercises.sort((a, b) => a.order - b.order);
    workouts.push({
      id: key.toLowerCase().replace(/\s+/g, "-"),
      programName: data.programName,
      week: data.week,
      dayOfWeek: data.day,
      exercises: data.exercises,
      isChecklist: data.day === "Daily",
    });
  }

  return { programs: programs.sort((a, b) => a.name.localeCompare(b.name)), workouts };
}
